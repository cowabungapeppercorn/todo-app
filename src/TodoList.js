import React from 'react';
import NewTodoForm from './NewTodoForm';
import Todo from './Todo';
import uuid from 'uuid/v4';

class TodoList extends React.Component {
  constructor(props){
    super(props);
    this.state = { todos: [] };
    this.addTodo = this.addTodo.bind(this);
    this.remove = this.remove.bind(this);
    this.editTodo = this.editTodo.bind(this);
    this.toggleEdit = this.toggleEdit.bind(this);
  }


  addTodo(todo) {
    let newTodo = { ...todo, id: uuid() };
    this.setState(state => ({
      todos: [...state.todos, newTodo]
    }));
  }

  remove(id) {
    let newTodoList = this.state.todos.filter(todo => todo.id !== id);
    this.setState({ todos: newTodoList });
  }

  toggleEdit(id) {
    let newTodoList = this.state.todos.map(todo => {
      if (todo.id === id) {
        todo.editMode = true;
        return todo;
      } else {
        return todo;
      }
    });
    this.setState({ todos: newTodoList });
  }

  editTodo(id, text) {
    let newTodoList = this.state.todos.map(todo => {
      if (todo.id === id) {
        todo.text = text;
        todo.editMode = false;
        return todo;
      } else {
        return todo;
      }
    });
    this.setState({ todos: newTodoList });
  }


  render() {
    return (
      <div>
        <NewTodoForm addTodo={this.addTodo} />
        <ul>
          {this.state.todos.map(todo => {
            return <Todo todoInfo={todo} toggleEdit={this.toggleEdit} editTodo={this.editTodo} key={todo.id} remove={this.remove}/>
          })}
        </ul>
      </div>
    );
  }
}

export default TodoList;