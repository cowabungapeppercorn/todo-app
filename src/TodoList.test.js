import React from 'react';
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json';
import TodoList from './TodoList';

it('renders without crashing', () => {
  shallow(<TodoList />);
});

it('renders as expected', () => {
  let wrapper = mount(<TodoList />);
  wrapper.setState({
    todos: [
      {
        text: "yo",
        id: 5,
        editMode: false
      }, {
        text: "gabba",
        id: 12335,
        editMode: false
      }
    ]
  });
  let serialized = toJson(wrapper);
  expect(serialized).toMatchSnapshot();
});

it('deletes a todo on delete button click', () => {
  let wrapper = mount(<TodoList />);
  wrapper.setState({
    todos: [
      {
        text: "yo",
        id: 5,
        editMode: false
      }, {
        text: "gabba",
        id: 12335,
        editMode: false
      }
    ]
  });
  wrapper.find('button').first().simulate('click');
  expect(wrapper.state().todos.length).toEqual(1);
});

it('renders edit todo form on edit button click', () => {
  let wrapper = mount(<TodoList />);
  wrapper.setState({
    todos: [
      {
        text: "yo",
        id: 5,
        editMode: false
      }, {
        text: "gabba",
        id: 12335,
        editMode: false
      }
    ]
  });
  wrapper.find('button').at(1).simulate('click');
  expect(wrapper.html()).toContain('editing...');
});

it('renders edit todo form on edit button click', () => {
  let wrapper = mount(<TodoList />);
  wrapper.setState({
    todos: [
      {
        text: "yo",
        id: 5,
        editMode: true
      }
    ]
  });
  let editTodoInput = wrapper.find('input').at(2);
  editTodoInput.instance().value = 'flinstones';
  editTodoInput.simulate('change');
  let submitBtn = wrapper.find('button').first();
  submitBtn.simulate('submit');
  expect(wrapper.state().todos[0].text).toEqual('flinstones');
});