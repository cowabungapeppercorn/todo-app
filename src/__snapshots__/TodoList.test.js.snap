// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<TodoList>
  <div>
    <NewTodoForm
      addTodo={[Function]}
    >
      <div>
        <form
          onSubmit={[Function]}
        >
          <label
            htmlFor="text"
          >
            Todo:
          </label>
          <input
            id="text"
            name="text"
            onChange={[Function]}
            value=""
          />
          <input
            type="submit"
          />
        </form>
      </div>
    </NewTodoForm>
    <ul>
      <Todo
        editTodo={[Function]}
        key="5"
        remove={[Function]}
        todoInfo={
          Object {
            "editMode": false,
            "id": 5,
            "text": "yo",
          }
        }
        toggleEdit={[Function]}
      >
        <li>
          yo
          <button
            onClick={[Function]}
          >
            X
          </button>
          <button
            onClick={[Function]}
          >
            Edit
          </button>
        </li>
      </Todo>
      <Todo
        editTodo={[Function]}
        key="12335"
        remove={[Function]}
        todoInfo={
          Object {
            "editMode": false,
            "id": 12335,
            "text": "gabba",
          }
        }
        toggleEdit={[Function]}
      >
        <li>
          gabba
          <button
            onClick={[Function]}
          >
            X
          </button>
          <button
            onClick={[Function]}
          >
            Edit
          </button>
        </li>
      </Todo>
    </ul>
  </div>
</TodoList>
`;
